jmclient/jmclient/wallet.py

```diff
@@ -688,14 +696,33 @@ def select_utxos(self, mixdepth, amount, utxo_filter=None,
             assert len(i) == 2
             assert isinstance(i[0], bytes)
             assert isinstance(i[1], numbers.Integral)
-        ret = self._utxos.select_utxos(
+        utxos = self._utxos.select_utxos(
             mixdepth, amount, utxo_filter, select_fn, maxheight=maxheight)
 
-        for data in ret.values():
+        total_value = 0
+        native_utxo = None
+        for key, data in utxos.items():
+            if self._is_native_utxo(data['path']):
+                native_utxo = key
+            total_value += data['value']
             data['script'] = self.get_script_from_path(data['path'])
             if includeaddr:
                 data["address"] = self.get_address_from_path(data["path"])
-        return ret
+
+        if require_auth_address and not native_utxo:
+            # try to select more utxos, hoping for a native one
+            try:
+                return self.select_utxos(
+                    mixdepth, total_value + 1, utxo_filter, select_fn,
+                    maxheight, includeaddr, require_auth_address)
+            except NotEnoughFundsException:
+                # recursive utxo selection was unsuccessful, give up
+                return {}
```

Comment:
If I'm reading this right, the current implementation just crashes if we get NotEnoughFunds, which bad as that may be, is not functionally the same as returning `{}`. So we need to check the outcome of that?

Reply:
The `NotEnoughFundsException` should not happen with the current implementation. But since we try to request a higher amount (to get an additional utxo here) it's possible that the exception can be raised.

Reply:
My concern is: suppose we reach this line of code and return `{}` as the selection procedure has failed due to this ioauth problem (or related), is it going to cause a problem in the caller which currently assumed that an exception would be raised?

See `jmclient.taker.Taker.prepare_my_bitcoin_data`, an `Exception` is caught (yes I know that's bad but anyway), triggering an `ABORT` callback.

The philosophical/design question is, if we can't satisfy the request, how should it be signalled to callers? The `NotEnoughFundsException` seems to make more sense to me than returning an empty dict?

Reply:
The caller did not assume any exception to have been raised and the exception would still be raised if the caller erroneously requests an amount that cannot be satisfied by the specified mixdepth. This empty (False-ish) return value is only used if the new parameter `require_auth_address` is explicitly set to `True`, which is only the case in the caller added in this PR which handles the returned empty dict correctly.

Reply:
OK, checked how it's called and yes there's no possibility of misuse as per your point of defaults and where the code is changed. Fine with this as-is.